Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/locate.h

```diff
@@ -0,0 +1,1886 @@
+// Copyright (c) 2014, 2017, 2018 INRIA (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+// You can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// Licensees holding a valid commercial license may use this file in
+// accordance with the commercial license agreement provided with the software.
+//
+// This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+// WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0+
+//
+// Author(s)     : Mael Rouxel-Labbé,
+//                 Stephen Kiazyk
+//
+#ifndef CGAL_POLYGON_MESH_PROCESSING_LOCATE_H
+#define CGAL_POLYGON_MESH_PROCESSING_LOCATE_H
+
+#include <CGAL/license/Polygon_mesh_processing/locate.h>
+
+#include <CGAL/AABB_face_graph_triangle_primitive.h>
+#include <CGAL/AABB_traits.h>
+#include <CGAL/AABB_tree.h>
+#include <CGAL/boost/graph/helpers.h>
+#include <CGAL/boost/graph/iterator.h>
+#include <CGAL/Default.h>
+#include <CGAL/Kernel_traits.h>
+#include <CGAL/Kernel/global_functions.h>
+#include <CGAL/Polygon_mesh_processing/internal/named_params_helper.h>
+#include <CGAL/Random.h>
+
+#include <boost/graph/graph_traits.hpp>
+#include <boost/mpl/if.hpp>
+#include <boost/variant.hpp>
+
+#include <array>
+#include <iostream>
+#include <iterator>
+#include <limits>
+#include <set>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+// Everywhere in this file:
+// If `tm` is the input triangulated surface mesh and given the pair (`f`, `bc`)
+// such that `bc` is `(w0, w1, w2)`, the correspondance with the weights in `bc`
+// and the vertices of the face `f` is the following:
+// - `w0` corresponds to `source(halfedge(f, tm), tm)`
+// - `w1` corresponds to `target(halfedge(f, tm), tm)`
+// - `w2` corresponds to `target(next(halfedge(f, tm), tm), tm)`
+
+namespace CGAL {
+
+namespace Polygon_mesh_processing {
+
+namespace internal {
+
+// to not carry 5-lines-long types everywhere
+template <typename TriangleMesh,
+          typename NamedParameters = Default>
+struct Locate_types
+{
+  typedef typename boost::graph_traits<TriangleMesh>::vertex_descriptor    vertex_descriptor;
+  typedef typename boost::graph_traits<TriangleMesh>::halfedge_descriptor  halfedge_descriptor;
+  typedef typename boost::graph_traits<TriangleMesh>::face_descriptor      face_descriptor;
+  typedef boost::variant<vertex_descriptor,
+                         halfedge_descriptor,
+                         face_descriptor>                                  descriptor_variant;
+
+  typedef typename std::conditional<
```

Comment:
Why not directly `GetVertexPointMap`?

Reply:
![](https://i.pinimg.com/originals/6d/0c/88/6d0c8841dd29ea3b3eb3625e96846b5d.jpg)